File: uDemy/SearchAlgorithms/binarySearch.js
// Binary Search
// Write a function called binarySearch which accepts a sorted array and
// a value and returns the index at which the value exists. Otherwise, return -1.

// Time Complexity - O(log n)

function binarySearch(arr, val) {
  let min = 0;
  let max = arr.length-1;

  while (min <= max) {
    let mid = Math.floor((max+min)/2);
    console.log(min,max,mid)
    if (val > arr[mid]) min = mid+1;
    else if (val < arr[mid]) max = mid-1;
    else return val;
  }

  return -1
}

console.log(binarySearch([1,2,3,4,5,6,7], 6)) // 6
console.log(binarySearch([2,4,6,8,10], 3)) // -`

File: LeetCode/arrays/binarySearch.py
# Given a sorted (in ascending order) integer array nums of n elements and a target value, write a function to search target in nums. If target exists, then return its index, otherwise return -1.


# Example 1:

# Input: nums = [-1,0,3,5,9,12], target = 9
# Output: 4
# Explanation: 9 exists in nums and its index is 4

# Example 2:

# Input: nums = [-1,0,3,5,9,12], target = 2
# Output: -1
# Explanation: 2 does not exist in nums so return -1

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        start = 0
        end = len(nums)-1
        while start < end:
            mid = floor((start+end)//2)
            if nums[mid] == target:
                return mid
            if target > nums[mid]:
                start = mid+1
            else:
                end = mid-1
        return start if nums[start] == target else -1
                